### DISCUSSION TITLE
Welcome to Synapse Discussions!

### DISCUSSION THREAD

**Discussion by geby** [geby/synapse#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by UralAlex** [comment]
Good afternoon! My name is Alex.
I'm learning the Delphi language, and in particular I need to write a small program to communicate with a server using SSH.
There are no ready-made built-in libraries and components in Delphi, unfortunately, I came across your Synapse library on the Internet, this is just my option,
I need to send remote commands to the server and get the result.
I downloaded this library and added it to my project, but I encountered a problem: my program does not compile, but produces an error.
He doesn't like the SSLCheck function in the ssl_cryptlib file, here's the message from Delphi:
[dcc32 Error] ssl_cryptlib.pas(227): E2026 Constant expression expected
[dcc32 Error] ssl_cryptlib.pas(676): E2003 Undeclared identifier: 'CRYPTLIB_VERSION'
[dcc32 Fatal Error] KillSession_s.pas(7): F2063 Could not compile used unit 'ssl_cryptlib.pas'
Having searched the Internet for this error, I came across an article that I did not have the CRYPTLIB_VERSION variable declared, 
I added this variable to the function var CRYPTLIB_VERSION: integer;
After this the error remained the same. Please help me please! Maybe we need to declare this variable somewhere else, or maybe as a constant?

**Reply by geby** [reply]
CRYPTLIB_VERSION is the constant declared in cryptlib.pas. (You can find it in the official Cryptlib ZIP download - you need the cl32.dll and cryptlib.pas unit.) Ssl_cryptlib.pas have dependency on cryptlib unit in the USES.

However cryptlib code is very old and may be it need revision. You can bypass old code and try this:

```
function TSSLCryptLib.SSLCheck(Value: integer): Boolean;
begin
  Result := true;
  FLastErrorDesc := '';
  if Value = CRYPT_ERROR_COMPLETE then
    Value := 0;
  FLastError := Value;
  if FLastError <> 0 then
  begin
    Result := False;
    FLastErrorDesc := GetString(FCryptSession, CRYPT_ATTRIBUTE_ERRORMESSAGE);
  end;
end;
```

**Comment by UralAlex** [comment]
Thanks! The error has disappeared, but for some reason it does not work to connect, it does not issue errors. If I attach my code here, maybe as soon as you have free time, can you tell me what's wrong, maybe I'm pointing something wrong? Thank you in advance! 
``` unit sshclient;

interface

uses
 tlntsend, ssl_openssl, ssl_openssl_lib, ssl_cryptlib;

type
 TSSHClient = class
 private
 FTelnetSend: TTelnetSend;
 public
 constructor Create(AHost, APort, AUser, APass: string);
 destructor Destroy; override;
 procedure SendCommand(ACommand: string);
 procedure LogOut;
 function ReceiveData: string;
 function LogIn: Boolean;
 end;

implementation

{ TSSHClient }

constructor TSSHClient.Create(AHost, APort, AUser, APass: string);
begin
 FTelnetSend := TTelnetSend.Create;
 FTelnetSend.TargetHost := AHost;
 FTelnetSend.TargetPort := APort;
 FTelnetSend.UserName := AUser;
 FTelnetSend.Password := APass;
end;

destructor TSSHClient.Destroy;
begin
 FTelnetSend.Free;
 inherited;
end;

function TSSHClient.LogIn: Boolean;
begin
 Result := FTelnetSend.SSHLogin;
end;

procedure TSSHClient.LogOut;
begin
 FTelnetSend.Logout;
end;

function TSSHClient.ReceiveData: string;
var
 lPos: Integer;
begin
 Result := '';
 lPos := 1;
 while FTelnetSend.Sock.CanRead(1000) or (FTelnetSend.Sock.WaitingData>0) do
  begin
    FTelnetSend.Sock.RecvPacket(1000);
    Result := Result + Copy(FTelnetSend.SessionLog, lPos, Length(FTelnetSend.SessionLog));
    lPos := Length(FTelnetSend.SessionLog)+1;
  end;
end;

procedure TSSHClient.SendCommand(ACommand: string);
begin
 FTelnetSend.Send(ACommand + #13);
end;
end.

**Comment by UralAlex** [comment]
this is the code that I write in the button handler
```procedure TKillSession.Button1Click(Sender: TObject);
var
  lSSh: TSSHClient;
begin
  lSSh := TSSHClient.Create('[Hostname]', '[Port]', '[Login]', '[Password]');//(HostEdit.Text, PortEdit.Text, LogEdit.Text, PasEdit.Text);
  if lSSh.LogIn then
    begin
      Memo1.Lines.Add('Connected!.');
      (* Get welcome message *)
      {Memo1.Lines.Add(lSSh.ReceiveData);
      (* Send command *)
      lSSh.SendCommand('-h');
      (* Receive results *)
      Memo1.Lines.Add(lSSh.ReceiveData);
      lSSh.LogOut;
      Memo1.Lines.Add('Logged out.');}
    end
  else
    //Writeln('Can''t connect.');
    Memo1.Lines.Add('Can''t connect.');
    lSSh.Free;
end;
